Added filtering step to ivar.py

TFerrel8 opened the pull request:
Adds filtering step based on fastp to ivar pipeline
Included a filtering step that trims a few NT at the beginning and end of each read (globally)

TFerrel8 committed:
Add files via upload

Added fixed start/end trim feature for each read.
Lines 15,16, 79, 80, and 98 relate to this new feature

TFerrel8 committed:
Rename ivar.py to ivar_.py

Remove notes from ivar and added a " ' " to the end of trim_f_t.

TFerrel8 committed:
Added filtering step to ivar pipeline

Adds filtering step based on fastp to ivar pipeline
The filtering step trims the beginning and end of the read (globally)

TFerrel8 committed:
Delete ivar_.py

khyox commented on mappgene/subscripts/ivar.py:
```diff
@@ -12,6 +12,8 @@ def run_ivar(params):
     variant_frequency = params['variant_frequency']
     read_cutoff_bp = params['read_cutoff_bp']
+    thread_usage = params['threads']
+    trim_front_tail = params['trim_f_t']
```
I think it would help to have the parameter with the same name, so 'trim_front_tail' too.

khyox commented on mappgene/subscripts/ivar.py:
```diff
@@ -70,6 +72,8 @@ def run_ivar(params):
 
     
     align_prefix = join(alignments_dir, subject)
+    r1_fastp = join(output_dir, '_R1_fastp_temp.fastq')
```
The name of the file would be missing with this set up. Maybe we can try:
```
r1_fastp = replace_extension(read1, '_R1_fastp_temp.fastq')
```

khyox commented on mappgene/subscripts/ivar.py:
```diff
@@ -70,6 +72,8 @@ def run_ivar(params):
     
     align_prefix = join(alignments_dir, subject)
+    r1_fastp = join(output_dir, '_R1_fastp_temp.fastq')
+    r2_fastp = join(output_dir, '_R2_fastp_temp.fastq')
```
Same as previous line, so perhaphs:
```
r2_fastp = replace_extension(read2, '_R2_fastp_temp.fastq')
```

khyox commented on mappgene/subscripts/ivar.py:
```diff
@@ -87,6 +91,7 @@ def run_ivar(params):
     output_fa = join(output_dir, f'{subject}.ivar.consensus')
     run(f'bwa index {fasta}', params)
+    run(f'fastp -i {read1} -I {read2} -o {r1_fastp} -O {r2_fastp} -D --trim_front1 {trim_front_tail} --trim_tail1 {trim_front_tail} -l 25 -w {thread_usage} -h {Subject_dir}'.fastp.html' 2>dev/null}', params) 
     run(f'bwa mem -t 8 {fasta} {read1} {read2} | samtools sort -o {bam}', params)
```
We have to update this line to make sure that `bwa mem` gets the filtered input from `fastp` and not the "raw" fastq files as it does now.

khyox reviewed (changes requested):
Thanks, much better now! 
I let some in-line comments in the code for some changes. Once we finish with `ivar.py`, we'll have to tare of `cli.py` script to add the new parameters there.

TFerrel8 committed:
Update ivar.py

Changed variable names and the input for bwa mem.
Lines 15-16: Changed variable names to match the parameter name.
Lines 75-76: Changed joined to replace_extension.
Line 94: Changed threads_usage to threads
Line 95: Changed read1/read2 to r1_fastp/r2_fastp

khyox reviewed (changes requested):
Perfect! I think `ivar.py` is right now and we can focus on the new required parameters: we need to make sure that those are populated so that `ivar.py` will work. So, it's time to turn our attention to [cli.py](https://github.com/khyox/mappgene/blob/01bcce1ec0ed7ae13dcb40b3b98a037c26b05328/mappgene/cli.py) to:
1.  Add the new arguments to the parser (see examples in lines 18–63 and, if needed, detailed documentation [here](https://docs.python.org/3.8/library/argparse.html).
2. Add the parser's output for the new arguments to the `base_params` dictionary (see line 73 of the current code)

(Remember to directly modify `cli.py` since git will take care of versioning)
Thanks!

TFerrel8 committed:
Added new parameters to cli.py

Added fastp parameters that relate to read trimming
Lines 21-25: 'trim_front_tail' and 'threads' parameters were added

khyox commented on mappgene/cli.py:
```diff
@@ -17,6 +17,12 @@ def parse_args(args):
         help='Test using the example inputs.')
+        
+    parser.add_argument('--threads' default=4, 
+        help='Allows the script to operate more efficiently)
```
That's totally true, but let's suppose that the user already knows this and alternatively we can say something more informative to our particular case such as "Number of threads used by the fastp filtering step".

khyox commented on mappgene/cli.py:
```diff
@@ -78,6 +84,8 @@ def main():
         'read_cutoff_bp': args.read_cutoff_bp,
         'primers_bp': args.primers_bp,
         'depth_cap': float(args.depth_cap),
+        'threads':int(args.threads),
```
Add missing space after `:` (PEP8)

khyox commented on mappgene/cli.py:
```diff
@@ -17,6 +17,12 @@ def parse_args(args):
     parser.add_argument('--test', action='store_true',
         help='Test using the example inputs.')
+        
+    parser.add_argument('--threads' default=4, 
```
Missing `,` between argument name and `default`

khyox commented on mappgene/cli.py:
```diff
@@ -17,6 +17,12 @@ def parse_args(args):
+    parser.add_argument('--threads' default=4, 
+        help='Allows the script to operate more efficiently)
+        
+    parser.add_argument('--trim_front_tail' default=10,
```
Missing `,` between argument name and `default`

khyox reviewed (changes requested):
I think we've almost converged to the solution. Just some minor changes still remain. Please see the comments for details. Many thanks!

TFerrel8 committed:
Update cli.py

khyox reviewed (approved):
Congrats! Now, time for testing the upgraded mappgene pipeline. :)
